source/extensions/filters/common/expr/context.h

```diff
@@ -99,9 +158,25 @@ template <class T> class HeadersWrapper : public google::api::expr::runtime::Cel
     return convertHeaderEntry(
         arena_, Http::HeaderUtility::getAllOfHeaderAsString(*value_, Http::LowerCaseString(str)));
   }
-  int size() const override { return value_ == nullptr ? 0 : value_->size(); }
+  int size() const override { return ListKeys()->size(); }
```

Comment:
Not an expert in this area and this may be a premature optimization, but how many times `size()` or `ListKeys()` are being called and should the returned value of `ListKeys()` be cached?

Reply:
I think we also need cache invalidation then, which is not supported right now. These wrappers are just views, like string_view, so caching requires some trigger to know when backing data changes. For example, if an expression is evaluated at the beginning of a request, it may accidentally cache 0 as response header size.

There is a better way to handle per-evaluation caching with memoization, since all functions are pure.